### PR TITLE
Detect and cleanly exit from renderer crashes

### LINKED ISSUE #6170: Detect browser/renderer crashes
### Current behavior:

Cypress doesn't know if the browser tab crashes during a test.

### Desired behavior:

Cypress detects browser tab crashes/browser crashes and exits cleanly after printing some debug information.

Dependency of: #349 

### Issues that this would make easier to debug

* Random hanging in Docker: https://github.com/cypress-io/cypress/issues/8206

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #6170 

### User facing changelog

- Cypress will now detect if your browser tab has crashed and will cleanly exit.

### Additional details

- previously, Cypress would hang open until the CI job timed out
- methods used to tell if the browser tab has crashed:
	- if the user is not `cy.visit`ing, and the socket disconnects, and there is not a new runner connection within 5 seconds, the browser tab crashed
	- if the user is `cy.visit`ing, the socket disconnects, and it does not reconnect within 10 seconds (generous)
	- [x] ~~this makes the code a lot more complex, maybe the rules should just be "if the runner disconnects, and it does not reconnect within 10 seconds, the browser tab crashed" - don't try to be smart about it~~
		- apparently socket.io can and will disconnect and reconnect, so there should be a delay in either case
		- since 10 seconds is kinda a long delay, maybe these two cases should remain separate
- what do we do when an unexpected exit does occur?
	- browser is left open for debugging
	- in open mode: a warning is logged to stdio
	- in run mode: an error is thrown and tests end
- [x] if the extension d/c's in run mode, force exit
- [x] keep extension d/c error message the same

Most of the meat of the error has been moved to the docs. Docs PR: cypress-io/cypress-documentation#2413

## User Experience Changes

- users will now see the `BROWSER_CRASHED` error message in stdout if the browser crashes in open mode
- users will see their tests exit early if the browser crashes, or if the extension disconnects, in run mode
- webextension disconnections now cause tests to end in run mode

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? cypress-io/cypress-documentation#2413

### CODE REVIEW

**Comment by flotwig** [comment]
WIP, using the existing RENDERER_CRASHED error and cleaning it up a bit to be generic:

image

In run mode, it will just throw and exit.

File: packages/server/lib/errors.coffee [hunk 1]
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
 
     when "RENDERER_CRASHED"
       """
-      We detected that the Chromium Renderer process just crashed.
+      We detected that the #{arg1} renderer process running your tests just exited unexpectedly.
```
Comment (flotwig) [inline packages/server/lib/errors.coffee]:
```suggestion
      We detected that the #{arg1} browser process running your tests just exited unexpectedly.
```

File: packages/server/lib/errors.coffee [hunk 2]
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
 
     when "RENDERER_CRASHED"
```
Comment (flotwig) [inline packages/server/lib/errors.coffee]:
```suggestion
    when "BROWSER_CRASHED"
```

File: packages/server/lib/errors.coffee [hunk 3]
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
...
-      - You are running Docker (there is an easy fix for this: see link below)
-      - You are running lots of tests on a memory intense application
-      - You are running in a memory starved VM environment
-      - There are problems with your GPU / GPU drivers
-      - There are browser bugs in Chromium
-
-      You can learn more including how to fix Docker here:
-
-      https://on.cypress.io/renderer-process-crashed
+      Learn more at https://on.cypress.io/renderer-process-crashed
```
Comment (flotwig) [inline packages/server/lib/errors.coffee]:
```suggestion
      Learn more at https://on.cypress.io/browser-process-crashed
```

File: packages/server/lib/project.js [hunk 4]
```diff
@@ -368,6 +368,18 @@ class Project extends EE {
           })
         }
       },
+
+      onUnexpectedDisconnect: () => {
+        const err = errors.get('RENDERER_CRASHED', this.browser.displayName)
+
+        browsers.close(true)
```
Comment (flotwig) [inline packages/server/lib/project.js]: just keep it open for debugging

**Comment by chrisbreiding** [comment]
> what do we do when an unexpected exit does occur?

>    browser is left open for debugging
>    in open mode: a warning is logged to stdio

I tried this out and I'm not sure this is the best experience for open mode for the following reasons:

- The browser is completely frozen, so I can't do anything with it. 
- I tend not to look at stdio when in open mode, so I wouldn't have seen the warning if I didn't know it was going to be there. 
- If in global mode, there is no stdio. 

I think it would be better to close the browser and show the error in the desktop GUI.

**Comment by flotwig** [comment]
@chrisbreiding It was originally displayed as a warning in the desktop-gui, but since users will see this error frequently (for example, any time they manually exit the browser during tests), it was deemed too noisy.

**Comment by chrisbreiding** [comment]
I see. I didn't realize this would happen any time the browser closed.

**Comment by flotwig** [comment]
@chrisbreiding with that being said, do you still think it's better to close the browser? The only thing with that is that it could make it difficult to debug the source of websocket disconnections, since the browser would be gone.

**Comment by chrisbreiding** [comment]
I didn't realize this was related to websocket disconnects too. I thought it was just the browser freezing or crashing, in which case it seems like the browser isn't accessible for debugging. I guess if there's a way to distinguish, we could handle those cases differently. If not, erring on the side of leaving the browser open is probably preferrable.

**Review by brian-mann — COMMENTED** [review]
I'm going to need to pair with you 1:1 before approving this. There are some experience related things I need to verify. We can do this early next week.

**Pull request closed without merging**